src/s3/commands/uploadFile.ts

```diff
@@ -26,51 +38,103 @@ export interface FileSizeBytes {
- * Shows the output channel with "upload started" message.
- * Uploads the file (showing a progress bar).
- * Shows the output channel with "upload completed" message.
- * Refreshes the node.
- *
- * Node that the node is reset to displaying its first page of results.
- * The file that is uploaded won't necessary fall on the first page.
- * The user may need to load more pages to see the uploaded file reflected in the tree.
+ * Wizard to upload a file.
+ * If the S3BucketNode | S3FolderNode is given, it asks for the file then uploads to node
+ * If the node is not give, it asks for the file and also for a bucket, then uploads
+ * 
+ * @param s3Client - account to upload the file to
+ * @param node - bucket or folder to upload the file to
+ * @param document - open document (to use as default), if any
+ * 
  */
-export async function uploadFileCommand(
-    node: S3BucketNode | S3FolderNode,
+ export async function uploadFileCommand(
+    s3Client: S3Client,
+    node?: S3BucketNode | S3FolderNode,
+    document?: vscode.Uri,
     fileSizeBytes: FileSizeBytes = statFile,
+    getBucket = promptUserForBucket,
+    getFile = getFileToUpload,
     window = Window.vscode(),
-    commands = Commands.vscode(),
-    outputChannel = ext.outputChannel
-): Promise<void> {
-    getLogger().debug('UploadFile called for %O', node)
-
-    const fileLocation = await promptForFileLocation(window)
-    if (!fileLocation) {
-        getLogger().info('UploadFile cancelled')
-        telemetry.recordS3UploadObject({ result: 'Cancelled' })
-        return
+    outputChannel = ext.outputChannel,
+    commands = Commands.vscode()
+): Promise<void>{
+    let key: string
+    let bucket: S3.Bucket
+    let file: vscode.Uri | undefined
+    
+    if (node) {
+        file = await getFile(undefined, window)
+            if (!file) {
+                showOutputMessage('No file selected, cancelling upload', outputChannel)
+                getLogger().info('UploadFile cancelled')
+                telemetry.recordS3UploadObject({ result: 'Cancelled' })
+                return
+            }
+        key = node.path + path.basename(file.fsPath)
+        bucket = { Name: node.bucket.name }
+    } else {
+        while (true) {
+            file = await getFile(document, window)
+            if (file) {
+                try{
+                    const bucketResponse = await getBucket(s3Client)
+                    if (bucketResponse === 'back'){
+                        continue
+                    }
+                    if(bucketResponse == 'cancel'){
+                        showOutputMessage('No bucket selected, cancelling upload', outputChannel)
+                        getLogger().info('No bucket selected, cancelling upload')
+                        telemetry.recordS3UploadObject({ result: 'Cancelled' })
+                        return
+                    }
+                    
+                    bucket = bucketResponse as S3.Bucket
```

Comment:
to guard against `getBucket()` changing in the future, it is worth making a defensive "type check" check here.

```suggestion
                    if (!(bucketResponse as any).Name) {
                      throw Error(`bucketResponse is not a S3.Bucket`)
                    }
                    bucket = bucketResponse as S3.Bucket
```